### ISSUE TITLE
Support "OCULUS_multiview" WebGL extension

### ISSUE THREAD

**Issue by MilllerTime** [mrdoob/three.js#20368]
Three.js supports the `"OVR_multiview2"` WebGL2 extension, which is enabled by default in some browsers. However that extension only works on non-anti-aliased contexts, which results in a very rough image. Further, the Oculus browser only supports it behind a flag.

Oculus has implemented their own version of the extension in the Oculus browser (`"OCULUS_multiview"`), and it supports multi-sample anti-aliasing as well. It appears Three.js doesn't support this one though.

Details and an example implementation is available here: https://developer.oculus.com/documentation/oculus-browser/browser-multiview/

It seems like this is a huge performance win for Three.js users, while still supporting anti-aliasing unlike Firefox Reality and other browsers. A quote from the page above:
> Often, a CPU usage reduction of 25% - 50% is possible.

Is there any interest in implementing support for this Oculus specific, but very useful extension?

**Comment by mrdoob** [comment]
Yes! That'd be great. Is `OCULUS_multiview` enabled by default?

We used to support `OVR_multiview2` but I decided to remove the code because the lack of antialias support.

**Comment by MilllerTime** [comment]
@mrdoob Yes, according to the Oculus Developer site, `OCULUS_multiview` is enabled by default in the Oculus Browser 🙂

**Comment by Artyom17** [comment]
@mrdoob 
Yeah, OCULUS_multiview is enabled by default (it can be disabled via chrome://flags though, for debugging purposes). You can see an example of it usage here: https://artyom17.github.io/webxr-samples/multi-layer-mv.html (source: https://github.com/Artyom17/webxr-samples/blob/master/multi-layer-mv.html).
There is intent to propose the extension to WebGL WG, however, there are several drawbacks at the moment which will delay the acceptance of the extension. And, of course, it will have a different name. But I can't provide any ETA for that.

**Comment by mrdoob** [comment]
Thanks @Artyom17!

I'm open for it 👍

**Comment by MilllerTime** [comment]
Very cool, glad to hear this! Also, I hadn't realized you removed the `OVR_multiview2` support, but that makes sense - things look pretty rough in-headset with no AA. Looking forward to seeing these great features combined and working via the new extension 😀

**Comment by danrossi** [comment]
I was wondering how this is going to better support performance and AA in quest.

**Comment by danrossi** [comment]
I've been tracking down the most up to date patch and there doesn't seem to be one. The suggested "current PR" was using the old OVR_multiview extension the non anti-aliased one. Its really needed to fix aliasing and performance on Quest. I have somebody complaining about aliasing as it needs frame buffer upscaling but then it causes performance issues. 

I tried merging branches but sourcetree made a mess of it so I hand picked changes into this branch. However I discovered how hacky it is.

I had to migrate to using WebGPURenderer for the node system as I have a custom MSDF text renderer I have upgraded to using the node system for shader customisations. I render scalable SVG icons converted to MSDF text for buttons this way. I just discovered there is a completely different WebXRManager and the WebGLBackend is different to WebGLRenderer.  

The custom shader patch for multiview needs to be built using the nodes system like the UniformNode. I see everything going to this new renderer and node system and WebXR WebGPU is in the work which I tested. So it should be built using the node system somehow also including the occulus extension. So I wasted time with the patch and should be looking at the new rendering pipeline method. 

To save too much work just support it with the WebGLBackend for now. WebGPU WebXR may not need the multiview extension or they may provide a different feature for the device ? The node customisation only gets activated for WebGLBackend and if the extension is available and a flag set. 

https://github.com/danrossi/three.js-multiview/tree/oculus-multiview

**Comment by danrossi** [comment]
It's tricky to pull off with nodes. And it still requires additions to the node builder among other areas. So while the position uniform and color node can be modified externally possibly, other stuff has to be added in. 

I successfully added an option to include the extension in the builder. Following the same design to enable a flag on the mesh object for multiview.

https://github.com/danrossi/three.js-multiview/blob/multiview-nodes/src/renderers/webgl-fallback/nodes/GLSLNodeBuilder.js#L993

Here is the most basic vertex shader required.  Layout attribute is hardcoded as `location=` in the node builder yet `layout (num_views = 2) ` is required. But not sure what num_views is needed for. It's not being called anywhere to get an attribute location. 

```
#version 300 es
#extension GL_OVR_multiview2 : require
precision mediump float;
layout (num_views = 2) in;
in vec4 inPos;
uniform mat4 u_viewMatrices[2];
void main() {
  gl_Position = u_viewMatrices[gl_ViewID_OVR] * inPos;
}
```

I have to figure out which of these need to be transformed with a define referencing the VIEW_ID. possibly `v_cameraProjectionMatrix` . And how to go about that using nodes. I'll try figuring out how to add defines first.

Looking at babylonjs the number of views is hardcoded to 2.  So an index of 2 can be referenced everywhere without a config  ?

```
v_cameraProjectionMatrices[2]
```

But would that work in non webxr or does the shader need to be different for webxr. 

```
#define VIEW_ID gl_ViewID_OVR;
uniform mat4 v_cameraProjectionMatrices[2];
#define v_cameraProjectionMatrix v_cameraProjectionMatrices[VIEW_ID]

Generated vertex example

```
nodeVarying3 = uv;
	positionLocal = position;
	modelViewMatrix = ( v_cameraViewMatrix * v_nodeUniform6 );
	v_positionView = ( modelViewMatrix * vec4( positionLocal, 1.0 ) ).xyz;
	v_modelViewProjection = ( v_cameraProjectionMatrix * vec4( v_positionView, 1.0 ) );

	// result
	gl_Position = v_modelViewProjection;
```

**Comment by danrossi** [comment]
Multiview support has been added but not for this oculus extension. I think OVR_multiview2 will replace the oculus_multiview extension. I'm still testing it.

**Comment by Mugen87** [comment]
> Multiview support has been added

Please note that this has been done only for `WebGPURenderer`.

**Comment by Mugen87** [comment]
Considering this issue as done. `WebGPURenderer` has multiview support. There are no plans to add support to `WebGLRenderer`.
